patroni/postgresql/__init__.py

```diff
@@ -407,6 +407,25 @@ def start(self, timeout=None, task=None, block_callbacks=False, role=None):
         self._pending_restart = False
 
         configuration = self.config.effective_configuration
+        if "unix_socket_directories" in configuration:
+            for d in configuration["unix_socket_directories"].split(","):
+                d = os.path.join(self._data_dir, d.strip())
+                if not os.path.isdir(d):
+                    try:
+                        os.makedirs(d)
+                    except FileExistsError:
+                        raise PatroniException(
+                                "'%s' is defined in unix_socket_directories, but it is not a directory".format(d)
+                                )
```

Comment:
Lines 410-419 and 420-428 are very similar.
It would make sense to make the separate method that tries to create the directory and use it from both places.

Reply:
Fixed